### ISSUE TITLE
Add numerically stable methods to torch.distributions (e.g. logcdf)

### ISSUE THREAD

**Issue by Balandat** [pytorch/pytorch#52973]
## 🚀 Feature
Expose additional methods on (mainly continuous) probability distributions. Examples are `logcdf`, `logsf` (survival function), hazard ratio, etc. 

## Motivation
These methods are useful for scientific computing and ML (e.g. information-based methods). While they can obviously be derived from some of the basic methods (`cdf(x).log()`), those computations are often less accurate and frequently result in numerical issues. Other packages expose many of these already: 
Scipy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html#scipy.stats.norm
Jax: https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.stats.norm
Tensorflow: https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Normal#log_cdf

## Pitch
Implement the most commonly used methods. `logcdf` and `logsf` would be a good start.  

## Alternatives
Use the base methods and somehow deal with numerical issues.

**Comment by Balandat** [comment]
@vishwakftw

**Comment by vishwakftw** [comment]
Let me prototype them quickly with the current API. Meanwhile cc: @fritzo  @neerajprad @alicanb (seems like I can't label issues anymore).

**Comment by vishwakftw** [comment]
This is also a duplicate of https://github.com/pytorch/pytorch/issues/32632.

@alicanb added a version of `logcdf` and `logsf` in this [gist[(https://gist.github.com/alicanb/8383e93aac3edb148168d029e5bdd20a) which seems to perform poorly for small x i.e., (x << -1). Let me check how to reduce the error in that subset of the domain.

**Comment by vishwakftw** [comment]
So after more investigation, I found out the following:
1. the `torch.distributions.Normal(0, 1).cdf` method is inaccurate because of issues with `erf` which is what the method is based on.
2. the results from the gist earlier were based on fp64 rather than fp32. The accuracy of the naive implementation is worsens quicker around `x \approx -4` in fp32.
3. I coded up a quick version of the approximation strategy in TFP / Jax (they are the same IIUC). The idea is to apply different piecewise approximates in different parts of the domain, and this seems to work reasonably well. The domain (real numbers) is split based of when the standard computations such as `log(cdf(x))` and `1 - log(cdf(-x))` becomes 0.
4. Because of the imprecision in the exisiting `cdf`, the active part of `log(cdf(x))` (i.e., when `0 < log(cdf(x)) < 1`) is smaller than the choices (which are conservative themselves) in TFP / Jax.
5. However, adjusting for this is causing problems with the approximations. Look at the difference plots below.

![image](https://user-images.githubusercontent.com/23639302/109434209-2ed1c000-79e2-11eb-8aad-148c9996c60b.png)

![image](https://user-images.githubusercontent.com/23639302/109434279-87a15880-79e2-11eb-815b-f77ca3be20ec.png)

So, ideally we will have to fix the `cdf` implementation (using Cephes like SciPy / TFP / Jax) to resolve the regions of inaccuracies. The defined active regions in the plots below are: (-3, 5) for fp32 and (-6, 8)

The plots were generated using 30000 points between -20 and 10. The code to reproduce the plots are given below (click **details**).

<details><summary> details </summary>
<p>

```
def log_ndtr_tfp_version(value: torch.Tensor):
    """
    Function to compute the log of the normal CDF at value.
    This is based on the TFP implementation.
    """
    dtype = value.dtype
    if dtype == torch.float64:
        lower, upper = -6, 8
    elif dtype == torch.float32:
        lower, upper = -3, 5
    else:
        raise TypeError("value needs to be either float32 or float64")
    
    # When x < lower, then we perform a fixed series expansion (asymptotic)
    # = log(cdf(x)) = log(1 - cdf(-x)) = log(1 / 2 * erfc(-x / sqrt(2))) = log(-1 / sqrt(2 * pi) * exp(-x ** 2 / 2) / x * (1 + sum))
    # When x >= lower and x <= upper, then we simply perform log(cdf(x))
    # When x > upper, then we use the approximation log(cdf(x)) = log(1 - cdf(-x)) \approx -cdf(-x)
    normal = torch.distributions.Normal(0, 1)
    return torch.where(value > upper,
                       torch.log1p(-normal.cdf(-value)),
                       torch.where(value >= lower,
                                   torch.log(normal.cdf(value)),
                                   log_ndtr_series_tfp(value)))

def log_ndtr_series_tfp(value: torch.Tensor, num_terms=3):
    """
    Function to compute the asymptotic series expansion of the log of normal CDF
    at value.
    This is based on the TFP implementation.
    """
    # sum = sum_{n=1}^{num_terms} (-1)^{n} (2n - 1)!! / x^{2n}))
    value_sq = value ** 2
    t1 = -0.5 * (math.log(2 * math.pi) + value_sq) - torch.log(-value)
    t2 = torch.zeros_like(value)
    value_even_power = value_sq.clone()
    double_fac = 1
    multiplier = -1
    for n in range(1, num_terms + 1):
        t2.add_(multiplier * double_fac / value_even_power)
        value_even_power.mul_(value_sq)
        double_fac *= (2 * n - 1)
        multiplier *= -1
    return t1 + torch.log1p(t2)

x = torch.linspace(-20, 10, 30000, dtype=torch.float32)
out = log_ndtr_tfp_version(x)
plt.plot(x.numpy(), abs(out.numpy() - log_ndtr(x.numpy())), label='abs(PyTorch - SciPy) (float32)')
plt.legend()
plt.show()

x = torch.linspace(-20, 10, 30000, dtype=torch.float64)
out = log_ndtr_tfp_version(x)
plt.plot(x.numpy(), abs(out.numpy() - log_ndtr(x.numpy())), label='abs(PyTorch - SciPy) (float64)')
plt.legend()
plt.show()
```
</p>
</details>

**Comment by vishwakftw** [comment]
With the SciPy implementation of `ndtr` in PyTorch for the standard normal CDF, I was able to reduce the error by several orders of magnitude. The plots below show this:

![image](https://user-images.githubusercontent.com/23639302/109443262-196f8c80-7a08-11eb-8547-1f1f3f8a0c77.png)

![image](https://user-images.githubusercontent.com/23639302/109443270-1eccd700-7a08-11eb-9164-d011167800ce.png)

The new implementation the `cdf` function is also more accurate than the existing version of `Normal(0, 1).cdf`.

![image](https://user-images.githubusercontent.com/23639302/109443735-4f614080-7a09-11eb-828e-adc726d53e0f.png)

![image](https://user-images.githubusercontent.com/23639302/109443750-5425f480-7a09-11eb-8a1e-facd2aba8640.png)

Perhaps, we should have a separate implementation of `ndtr` and `log_ndtr` in ATen; this could prevent extraneous computations (the `input`, `other` in `torch.where` have unused computations conditionally) and leverage potential compiler optimizations.

The code is appended below (click **details**).

<details> <summary> details </summary>
<p>

```python
def new_ndtr(value: torch.Tensor):
    """
    Based on the SciPy implementation of ndtr
    """
    sqrt_half = torch.sqrt(torch.tensor(0.5, dtype=value.dtype))
    x = value * sqrt_half
    z = abs(x)
    y = 0.5 * torch.erfc(z)
    output = torch.where(z < sqrt_half,
                         0.5 + 0.5 * torch.erf(x),
                         torch.where(x > 0, 1 - y, y))
    return output


def log_ndtr_tfp_version(value: torch.Tensor, ndtr_new: bool = False):
    """
    Function to compute the log of the normal CDF at value.
    This is based on the TFP implementation.
    """
    dtype = value.dtype
    if dtype == torch.float64:
        lower, upper = -20, 8
    elif dtype == torch.float32:
        lower, upper = -10, 5
    else:
        raise TypeError("value needs to be either float32 or float64")
    
    # When x < lower, then we perform a fixed series expansion (asymptotic)
    # = log(cdf(x)) = log(1 - cdf(-x)) = log(1 / 2 * erfc(-x / sqrt(2)))
    # = log(-1 / sqrt(2 * pi) * exp(-x ** 2 / 2) / x * (1 + sum))
    # When x >= lower and x <= upper, then we simply perform log(cdf(x))
    # When x > upper, then we use the approximation log(cdf(x)) = log(1 - cdf(-x)) \approx -cdf(-x)
    if ndtr_new:
        ndtr_func = new_ndtr
    else:
        ndtr_func = torch.distributions.Normal(0, 1).cdf
    return torch.where(value > upper,
                       torch.log1p(-ndtr_func(-value)),
                       torch.where(value >= lower,
                                   torch.log(ndtr_func(value)),
                                   log_ndtr_series_tfp(value)))

def log_ndtr_series_tfp(value: torch.Tensor, num_terms=3):
    """
    Function to compute the asymptotic series expansion of the log of normal CDF
    at value.
    This is based on the TFP implementation.
    """
    # sum = sum_{n=1}^{num_terms} (-1)^{n} (2n - 1)!! / x^{2n}))
    value_sq = value ** 2
    t1 = -0.5 * (math.log(2 * math.pi) + value_sq) - torch.log(-value)
    t2 = torch.zeros_like(value)
    value_even_power = value_sq.clone()
    double_fac = 1
    multiplier = -1
    for n in range(1, num_terms + 1):
        t2.add_(multiplier * double_fac / value_even_power)
        value_even_power.mul_(value_sq)
        double_fac *= (2 * n - 1)
        multiplier *= -1
    return t1 + torch.log1p(t2)

x = torch.linspace(-30, 10, 40000, dtype=torch.float32)
out = log_ndtr_tfp_version(x, ndtr_new=True)
plt.plot(x.numpy(), abs(out.numpy() - log_ndtr(x.numpy())), label='abs(PyTorch - SciPy) (float32)')
plt.legend()
plt.show()

x = torch.linspace(-30, 10, 40000, dtype=torch.float64)
out = log_ndtr_tfp_version(x, ndtr_new=True)
plt.plot(x.numpy(), abs(out.numpy() - log_ndtr(x.numpy())), label='abs(PyTorch - SciPy) (float64)')
plt.legend()
plt.show()

x = torch.linspace(-30, 10, 40000, dtype=torch.float32)
plt.plot(x.numpy(), abs(torch.distributions.Normal(0, 1).cdf(x).numpy() - ndtr(x.numpy())), label='Old')
plt.plot(x.numpy(), abs(new_ndtr(x).numpy() - ndtr(x.numpy())), label='New')
plt.title('Float32')
plt.legend()
plt.show()

x = torch.linspace(-30, 10, 40000, dtype=torch.float64)
plt.plot(x.numpy(), abs(torch.distributions.Normal(0, 1).cdf(x).numpy() - ndtr(x.numpy())), label='Old')
plt.plot(x.numpy(), abs(new_ndtr(x).numpy() - ndtr(x.numpy())), label='New')
plt.title('Float64')
plt.legend()
plt.show()
```
</p>
</details>

**Comment by 0x00b1** [comment]
A brief update to this issue: `ndtr` was added to `torch.special` in 1.10.0.

**Comment by Balandat** [comment]
cc @j-wilson

**Comment by shchur** [comment]
I have created a [small library](https://github.com/shchur/survival_distributions) that implements `logsf`, `logcdf`, `log_hazard` and `isf` for some univariate distributions in`torch.distributions`. The implementation is numerically stable (compared to naive `(1.0 - p.cdf(x)).log()`) for the distributions that don't require special functions, such as `Exponential`, `Weibull`, `Uniform`, `Logistic`, `LogLogistic`, `MixtureSameFamily`. However, for `Normal` and `LogNormal` it just relies on the naive implementation.

I will be happy to contribute this as a pull request, if this is something that you are willing to review.

**Comment by cdaube** [comment]
It seems someone has ported the tensorflow (log-)ndtr to torch:
https://github.com/probabll/mixed-rv-vae/blob/fe809beb42f3c4d0d388ccd534cdba800d4d0a72/torch_log_ndtr.py

As far as I can tell, this is a bit better behaved than the current ndtr in torch.
